Make GATKSVPipelineSingleSample validate it in MiniWDL

illusional opened the pull request:
Fixes #153.

Make the GATKSVPipelineSingleSample pipeline validate in MiniWDL through a combination of:

- Optional to non-optional
  - Eg: Output was required, but input `outfile-name` was optional.
  - Add `select_first` to coerce `X? -> X`
- Remove duplicate inputs
- Default disk_overhead_gb to 0: `disk_overhead_gb = select_first([disk_overhead_gb, 0])`, because if it's not supplied I guess it could be 0.

epiercehoffman commented:
Thank you for bringing to our attention that users of MiniWDL may encounter these validation issues. We use womtool for validation and cromwell to run workflows, and it appears that those tools have different validation checks. 

I believe that MiniWDL is not handling parameters with default values the same way as womtool/cromwell, based on the changes you made with regard to optional/required inputs. In the cases of `defragment_max_dist`, `min_large_pesr_call_size_for_filtering`, `min_large_pesr_depth_overlap_fraction`, and `disk_overhead_gb`, these inputs should be optional in the top-level WDL, as all of them have default values assigned at a task-level (in the input section of the task). Since they have default values, any WDL that calls these tasks is not required to provide a value for these parameters when running workflows with cromwell. We do not wish for users to have to provide inputs for parameters that already have default values, so these parameters should remain optional. 

With regard to your other changes, thank you for bringing to our attention the duplicate inputs, the lack of newline at the end of Tasks0506.wdl, and the use of the optional input `outfile_name` in `Tasks0506.ZcatCompressedFiles` when the task should be using `output_file_name` or assigning `outfile_name` a default value to deal with the optional/required problem.

illusional commented:
Thanks for the reply @epiercehoffman! 

I believe Cromwell / womtool is actually parsing this incorrectly with respect to the WDL spec. I've requested clarification here: https://github.com/openwdl/wdl/issues/462.

To address your feedback, I've made the Optional -> Non optional map apply the default in the task, like:

```wdl
task ... {
  input {
    Int? disk_overhead_gb = 10
  }
  disk_overhead_gb_ = select_first([disk_overhead_gb, 10])
  # ... use `disk_overhead_gb_`
}
```

This validates with womtool and MiniWDL.

illusional commented on wdl/GATKSVPipelineSingleSample.wdl:
```diff
@@ -339,7 +338,6 @@ workflow GATKSVPipelineSingleSample {
 
     # Depth part 2
     RuntimeAttr? runtime_attr_integrate_depth_gq
-    RuntimeAttr? runtime_attr_concat_vcfs
```
This is only used in module04, not passed to any other task.

mwalker174 replied:
Yes but in 04 there should be 2 separate `runtime_attr_concat_vcfs` inputs, one for `GenotypeDepthPart2` and one for `GenotypePESRPart2`.

illusional commented on wdl/GATKSVPipelineSingleSample.wdl:
```diff
@@ -364,7 +362,6 @@ workflow GATKSVPipelineSingleSample {
 
     RuntimeAttr? runtime_override_update_sr_list
     RuntimeAttr? runtime_override_merge_pesr_depth
-    RuntimeAttr? runtime_override_merge_pesr_depth
```
@mwalker, I don't know if this is a copy-paste error, or the `breakpoint_overlap_filter` is intentionally assigned from `merge_pesr_depth`.

mwalker174 replied:
This does look like an error, can you add a `runtime_override_breakpoint_overlap_filter` input?

illusional commented on wdl/Module00c.wdl:
```diff
@@ -288,7 +289,7 @@ workflow Module00c {
     input:
       samples = all_samples,
-      BAF_files = if defined(BAFFromShardedVCF.baf_files) then BAFFromShardedVCF.baf_files else BAF_files,
+      BAF_files = if defined(BAFFromShardedVCF.baf_files) then BAFFromShardedVCF.baf_files else (if defined(BAF_files) then select_all(select_first([BAF_files])) else NONE_ARRAY_FILE_),
```
- BAFFromShardedVCF.baf_files -> Array[File]?
- BAF_files -> Array[File?]?

Both of which are valid, but the spec disallows this:

>  The return type of the if-then-else should be the same, regardless of which side is evaluated or runtime problems might occur. [[source](https://github.com/openwdl/wdl/blob/main/versions/1.0/SPEC.md#:~:text=%20the%20return%20type%20of%20the%20if-then-else%20should%20be%20the%20same%2C%20regardless%20of%20which%20side%20is%20evaluated%20or%20runtime%20problems%20might%20occur.)]

Error from MiniWDL:

> Expected Array[File]? instead of Array[File?]?  (if consequent & alternative must have the same type)

mlin replied:
> (if consequent & alternative must have the same type)

That error message is from a relatively old version of miniwdl. I think upgrading would resolve this case where it wasn't "unifying" the optionality of the array item type. (which is different from https://github.com/openwdl/wdl/issues/462, where I think you found a corner of the spec that isn't totally clear)

illusional replied:
😓  https://github.com/broadinstitute/wdl-ide/blob/76a04051677347fb9014ff6ef9c491b197f89d4a/server/setup.py#L34

```
wdl-lsp 0.0.72 requires miniwdl==0.5.1
```

---

Can confirm, miniwdl==1.1.5 validates the original correctly, I've reverted this portion.

illusional commented on wdl/DepthPreprocessing.wdl:
```diff
@@ -167,7 +169,7 @@ task MergeSample {
     bedtools merge -i ~{sample_id}.bed -d 0 -c 4,5,6,7 -o distinct > ~{sample_id}.merged.bed
     /opt/sv-pipeline/00_preprocessing/scripts/defragment_cnvs.py \
-      --max-dist ~{max_dist} ~{sample_id}.merged.bed ~{sample_id}.merged.defrag.bed
+      --max-dist ~{max_dist_}} ~{sample_id}.merged.bed ~{sample_id}.merged.defrag.bed
```
Yeah I agree with this (and the pattern we take for our workflows) just wanted to keep something similar.

mwalker174 replied:
Ah does this work `~{if defined(max_dist) then max_dist else "0.25"}`?

illusional commented on wdl/GATKSVPipelineSingleSample.wdl:
```diff
@@ -316,7 +316,6 @@ workflow GATKSVPipelineSingleSample {
     File bin_exclude
 
     # Common
-    RuntimeAttr? runtime_attr_split_vcf
```
I've split those two RuntimeAttr. I've left comments on the review for the other two.

illusional commented on wdl/MakeBincovMatrix.wdl:
```diff
@@ -73,18 +73,19 @@ task SetBins {
 
   Float disk_scale_factor = 10.0
-  Int disk_gb = disk_overhead_gb + ceil(disk_scale_factor * size(count_file, "GiB"))
+  Int disk_gb = disk_overhead_gb_ + ceil(disk_scale_factor * size(count_file, "GiB"))
```
Consolidated

illusional commented on wdl/Module00a.wdl:
```diff
@@ -196,12 +196,12 @@ workflow Module00a {
+    Float? coverage_i = if defined(coverage) then select_first([coverage])[0] else NONE_FLOAT_
+    Float? pct_chimeras_i = if defined(pct_chimeras) then select_first([pct_chimeras])[0] else NONE_FLOAT_
+    Float? total_reads_i = if defined(total_reads) then select_first([total_reads])[0] else NONE_INT_
+    Int? pf_reads_improper_pairs_i = if defined(pf_reads_improper_pairs) then select_first([pf_reads_improper_pairs])[0] else NONE_INT_
```
Made modifications to Module00aBatch, and directed the originals to insert_size instead of insert_size_input: https://github.com/broadinstitute/gatk-sv/blob/946f00a9ac3c5dcd8f90b63650ad496c0c22fb87/wdl/GATKSVPipelineSingleSample.wdl#L453-L459

illusional commented:
Thanks @mwalker174, I've addressed your comments. I've also reverted the `select_first([...])` changes you suggested as it causes Cromwell to fail to validate (an error with Cromwell). I've seen this before, and this is a grammar issue with Cromwell, where I know valid workflows are failing. It's trying to access the attribute `3` on the object `0` (actually an integer).

```
Failed to import 'SingleSampleFiltering.wdl' (reason 1 of 1): ERROR: Unexpected symbol (line 247, col 117) when parsing 'e'.

Expected identifier, got "3".

        | coverageBed -a stdin -b ~{raw_dels} | awk '$NF < ~{select_first([min_large_pesr_depth_overlap_fraction, 0.3])} {print $4}' > large_dels_without_raw_depth_support.list
                                                                                                                    ^

$e = $e <=> :dot :identifier -> MemberAccess( value=$0, member=$2 )
```

illusional commented:
Hi @mwalker174 and @epiercehoffman, are there any blockers for this PR, or anything I need to get it ready for re-review?

mwalker174 commented on wdl/DepthPreprocessing.wdl:
```diff
@@ -141,7 +141,7 @@ task MergeSample {
     File gcnv
     Array[File] cnmops
-    Float max_dist = 0.25
+    Float? max_dist = 0.25
```
Change to just `Float? max_dist`? Assigning the value is redundant with the line below.

illusional replied:
Yep, actioned so resolving.

mwalker174 reviewed (commented):
Sorry for the delayed response - I still have a couple of suggestions here.

illusional commented:
Thanks @mwalker174, I've refactored the `select_first` to `if defined(var) then var else "default"`, the key being the default being in the string due to the cromwell grammar bug.

mwalker174 reviewed (approved):
Thanks @illusional for the updates and follow-through, looks good!
